ledger/participant-integration-api/src/test/lib/scala/platform/indexer/ha/EndlessReadService.scala

```diff
@@ -0,0 +1,222 @@
+
+  override def ledgerInitialConditions(): Source[LedgerInitialConditions, NotUsed] = {
+    logger.info("EndlessReadService.ledgerInitialConditions() called")
+    initialConditionCalls.incrementAndGet()
+    Source
+      .single(LedgerInitialConditions(ledgerId, configuration, recordTime(0)))
+      .via(killSwitch.flow)
+  }
+
+  /** Produces the following stream of updates:
+    *    1. a config change
+    *    1. a party allocation
+    *    1. a package upload
+    *    1. a transaction that creates a contract
+    *    1. a transaction that archives the previous contract
+    *
+    *    The last two items above repeat indefinitely
+    */
+  override def stateUpdates(beginAfter: Option[Offset]): Source[(Offset, Update), NotUsed] = {
+    logger.info(s"EndlessReadService.stateUpdates($beginAfter) called")
+    stateUpdatesCalls.incrementAndGet()
+    val startIndex: Int = beginAfter.map(index).getOrElse(1)
+    Source
+      .fromIterator(() => Iterator.from(startIndex))
+      .throttle(updatesPerSecond, 1.second)
+      .map {
+        case i @ 1 =>
+          offset(i) -> Update.ConfigurationChanged(
+            recordTime(i),
+            submissionId(i),
+            participantId,
+            configuration,
+          )
+        case i @ 2 =>
+          offset(i) -> Update.PartyAddedToParticipant(
+            party,
+            "Operator",
+            participantId,
+            recordTime(i),
+            Some(submissionId(i)),
+          )
+        case i @ 3 =>
+          offset(i) -> Update.PublicPackageUpload(
+            List(archive),
+            Some("Package"),
+            recordTime(i),
+            Some(submissionId(i)),
+          )
+        case i if i % 2 == 0 =>
+          offset(i) -> Update.TransactionAccepted(
+            optCompletionInfo = Some(completionInfo(i)),
+            transactionMeta = transactionMeta(i),
+            transaction = createTransaction(i),
+            transactionId = transactionId(i),
+            recordTime = recordTime(i),
+            divulgedContracts = List.empty,
+            blindingInfo = None,
+          )
+        case i =>
+          offset(i) -> Update.TransactionAccepted(
+            optCompletionInfo = Some(completionInfo(i)),
+            transactionMeta = transactionMeta(i),
+            transaction = exerciseTransaction(i),
+            transactionId = transactionId(i),
+            recordTime = recordTime(i),
+            divulgedContracts = List.empty,
+            blindingInfo = None,
+          )
+      }
+      .via(killSwitch.flow)
+  }
+
+  def abort(cause: Throwable): Unit = synchronized {
+    logger.info(s"EndlessReadService.abort() called")
+    aborted = true
+    killSwitch.abort(cause)
+  }
+
+  def reset(): Unit = synchronized {
+    logger.info(s"EndlessReadService.reset() called")
+    stateUpdatesCalls.set(0)
+    initialConditionCalls.set(0)
+    aborted = false
+    killSwitch = KillSwitches.shared("EndlessReadService")
```

Comment:
sholudn't we need to shutdown the previous killswitch as well?

Reply:
The intention is that if you want to abort previously started update streams, you should explicitly call `abort`. `reset` is to make sure all future operations behave as if called on a fresh `EndlessReadService`.